File: Exercise1/R/PFsample3.R
#####################################################
# PFsample3.R                                       #
#                                                   #
# Data Assimilation on Advection-Diffusion Equation #
#####################################################

# parameters
u0     <- 8.0    # advection velocity (true/unknown)
v0     <- 5.0    # diffusion coefficient (true/unknown)
tau20  <- 1.0    # variance of external force (true/unknown)
sigma2 <- 16.0   # variance of observation noise (true/known)

um     <- 5.0    # mean of prior of u
usd    <- 3.0    # standard deviation of prior of u
vm     <- 6.0    # mean of prior of u
vsd    <- 2.0    # standard deviation of prior of v
tau2m  <- 0.8    # mean of prior of u
tau2sd <- 0.5    # log-standard deviation of prior of tau2

# grid spacing
dt <- 0.5            # grid spacing in time
Nt <- 1200           # number of grids in time
t  <- 0:(Nt-1) * dt  # real coordinate in time
dx <- 10.0           # grid spacing in space
Nx <- 400            # number of grids in space
x  <- 0:(Nx-1) * dx  # real coordinate in space
xl <- c(2:Nx,1)      # vector letting 1:Nx shift leftward
xr <- c(Nx,1:(Nx-1)) # vector letting 1:Nx shift rightward
Np <- 1000           # number of particles

# function for simulation
simulation <- function( u, v, tau2 ) {
  C <- matrix( 0, nrow=Nx, ncol=Nt )

  for ( k in 2:Nt ) {
    A <- - u * ( C[xl,k-1] - C[xr,k-1] ) / ( 2 * dx )          # advection
    K <- max(k-2,1)
    S <- v * ( C[xl,K] - 2 * C[ ,K] + C[xr,K] ) / ( dx * dx )  # diffusion
    W <- sin( pi * x / 60 ) * sin( 2 * pi * t[k-1] / 120 ) +
           rnorm( Nx, mean=0, sd=sqrt(tau2) )                  # external force
    C[ ,k] <- 2 * dt * ( A + S + W )
    if ( k>2 ) C[ ,k] <- C[ ,k-2] + C[ ,k]
  }

  return( C )
}

# synthetic observation data
yt0 <- simulation( u0, v0, tau20 )  # true
yts <- simulation( um, vm, tau2m )  # simulation using guess parameters

yt  <- matrix( NA, nrow=Nx, ncol=Nt )  # observed concentration
observedX <- seq( 3,Nx,20)             # observed places
observedT <- seq(10,Nt,15)             # observed time
yt[ observedX, observedT ] <- yt0[ observedX, observedT ]
yt <- yt + rnorm( length(yt), mean=0, sd=sqrt(sigma2) )
                            # add observation noise

# variables for DA
xtp  <- array( 0, dim=c(Nx,2,Np) )     # predictive distribution
xtf  <- array( 0, dim=c(Nx,2,Np) )     # filter distribution
xtfm <- matrix( 0, nrow=Nx, ncol=Nt )  # mean of predictive distribution
llh  <- rep( 0, length=Np )            # log-likelihood
res  <- rep( 0, length=Np )            # number of resamples

# sample particles from priors
u    <- rnorm( Np, mean=um, sd=usd )
v    <- rnorm( Np, mean=vm, sd=vsd )
tau2 <- rlnorm( Np, meanlog=log(tau2m), sdlog=tau2sd )

uPrior <- u; vPrior <- v; tau2Prior <- tau2

# particle filter
for ( k in 2:Nt ) {
  # one-step ahead prediction
  A <- - ( xtf[xl,1, ] - xtf[xr,1, ] ) %*% diag(u) / ( 2 * dx )
  S <- ( xtf[xl,2, ] - 2 * xtf[ ,2, ] + xtf[xr,2, ] ) %*% diag(v) / ( dx * dx )
  W <- sin( pi * x / 60 ) * sin( 2 * pi * t[k] / 120 ) +
         matrix( rnorm( Nx*Np, mean=0, sd=sqrt(tau2) ), nrow=Nx, byrow=T )
  xtp[ ,1, ] <- xtf[ ,2, ] + 2 * dt * ( A + S + W )
  xtp[ ,2, ] <- xtf[ ,1, ]

  # skip filtering step if no observation
  if ( all( observedT!=k ) ) {
    xtf <- xtp
    xtfm[ ,k] <- apply( xtf[ ,1, ], 1, mean )
    next
  } 

  # log-likelihood at this time point
  dy  <- yt[observedX,k] - xtp[observedX,1, ]
  llh <- - 0.5 * ( log( 2 * pi * sigma2 ) + apply( dy*dy, 2, sum ) / sigma2 )

  # filtering using the residual systematic resampling
  weight <- exp( llh - max(llh) ) / sum( exp( llh - max(llh) ) )
  r <- runif(1) / Np
  uTemp <- u; vTemp <- v; tau2Temp <- tau2

  for ( p in 1:Np ) {
    res[p] <- floor( ( weight[p] - r ) * Np ) + 1
    r <- r + res[p] / Np - weight[p]
  }

  for ( p in which(res>0) ) {
    prange <- ( sum( res[p-1] ) + 1 ) : sum( res[p] )
    xtf[ , ,prange] <- xtp[ , ,p]
    u[prange] <- uTemp[p]
    v[prange] <- vTemp[p]
    tau2[prange] <- tau2Temp[p]
  }

  xtfm[ ,k] <- apply( xtf[ ,1, ], 1, mean )
}

# figures
# prior & posterior distributions of the model parameters
jpeg( file="paramDist.jpg", quality=100,
      width=1000, height=1000, pointsize=24 )
par( mfcol=c(3,2) )
hist( uPrior, breaks=seq( floor(min(uPrior)), ceiling(max(uPrior)), 0.5 ),
      main="p(u)", xlab="u", col="gray" )
hist( vPrior, breaks=seq( floor(min(vPrior)), ceiling(max(vPrior)), 0.5 ),
      main="p(v)", xlab="v", col="gray" )
hist( tau2Prior, breaks=seq( floor(min(tau2Prior)), ceiling(max(tau2Prior)), 0.1 ),
      main="p(tau2)", xlab="tau2", col="gray" )
hist( u, breaks=seq( floor(min(uPrior)), ceiling(max(uPrior)), 0.5 ),
      main="p(u|y)", xlab="u", col="gray" )
hist( v, breaks=seq( floor(min(vPrior)), ceiling(max(vPrior)), 0.5 ),
      main="p(v|y)", xlab="v", col="gray" )
hist( tau2, breaks=seq( floor(min(tau2Prior)), ceiling(max(tau2Prior)), 0.1 ),
      main="p(tau2|y)", xlab="tau2", col="gray" )
dev.off()

# true, simulated, observed & assimilated concentrations
jpeg( file="concentration.jpg", quality=100,
      width=1200, height=1000, pointsize=24 )
par( mfrow=c(2,2), mar=c(4,5,2,2) )
image( x, t, yt0, xlim=c(0,Nx*dx), ylim=c(0,Nt*dt),
       main="True", xlab="", ylab="t" )
image( x, t, yts, xlim=c(0,Nx*dx), ylim=c(0,Nt*dt),
       main="Simulation", xlab="", ylab="t" )
image( x, t, yt , xlim=c(0,Nx*dx), ylim=c(0,Nt*dt),
       main="Observation", xlab="x", ylab="t" )
image( x, t, xtfm, xlim=c(0,Nx*dx), ylim=c(0,Nt*dt),
       main="Assimilation", xlab="x", ylab="t" )
dev.off()

File: Exercise1/Python/PFsample3.py
# ---
# jupyter:
#   jupytext:
#     text_representation:
#       extension: .py
#       format_name: light
#       format_version: '1.5'
#       jupytext_version: 1.7.1
#   kernelspec:
#     display_name: Python 3
#     language: python
#     name: python3
# ---

# +
'''
    PFsample1.py
    Sample Program for Particle Filter
    3. Data Assimilation on Advection-Diffusion Equation
'''

# import modules
import random
import numpy as np
import matplotlib.pyplot as plt
from tqdm import tqdm

# fix seed
RANDOM_SEED = 0
random.seed(RANDOM_SEED)
np.random.seed(RANDOM_SEED)

# parameters
u0 = 8.0
v0 = 5.0
tau20 = 1.0
sigma2 = 16.0
um = 5.0
usd = 3.0
vm = 6.0
vsd = 2.0
tau2m = 0.8
tau2sd = 0.5

# grid spacing
dt = 0.5
Nt = 1200
t = np.arange(Nt) * dt
dx = 10.0
Nx = 400
x = np.arange(Nx) * dx
xl = np.array([(i+1)%Nx for i in range(Nx)])
xr = np.array([(i-1)%Nx for i in range(Nx)])
Np = 1000


# +
# function for simulation
def simulation(u, v, tau2):
    
    C = np.zeros([Nt,Nx])
    
    for k in range(1,Nt):
        A = -u * (C[k-1,xl] - C[k-1,xr]) / (2 * dx)
        K = max(k-2, 0)
        S = v * (C[K,xl] - 2 * C[K,:] + C[K,xr]) / (dx * dx)
        W = np.sin(2 * np.pi * t[k-1] / 120) * np.sin(np.pi * x / 60) + np.random.normal(0, np.sqrt(tau2), Nx)
        C[k] = 2 * dt * (A + S + W)
        if k>1:
            C[k] = C[k-2] + C[k]
    
    return C

# synthetic observation data
yt0 = simulation(u0, v0, tau20)
yts = simulation(um, vm, tau2m)
yt = np.zeros([Nt,Nx]) + np.nan
observedT = np.arange(9, Nt, 15)
observedX = np.arange(2, Nx, 20)
for t_index in observedT:
    for x_index in observedX:
        yt[t_index,x_index] = yt0[t_index,x_index]

yt = yt + np.random.normal(0, np.sqrt(sigma2), len(yt)).reshape(-1, 1)

# variables for DA
xtp = np.zeros([Np,2,Nx])
xtf = np.zeros([Np,2,Nx])
xtfm = np.zeros([Nt,Nx])
llh = np.zeros(Np)
res = np.zeros(Np, dtype=int)

# sample particles from priors
u = np.random.normal(um, usd, Np)
v = np.random.normal(vm, vsd, Np)
tau2 = np.random.lognormal(np.log(tau2m), tau2sd, Np)
uPrior = u.copy()
vPrior = v.copy()
tau2Prior = tau2.copy()
# -

# particle filter
for k in tqdm(range(1,Nt)):
    
    # one-step ahead prediction
    tau2_expand = np.array([])
    for _ in range(Nx):
        tau2_expand = np.append(tau2_expand, tau2)
    A = - np.diag(u) @ (xtf[:,0,xl] - xtf[:,0,xr]) / (2 * dx)
    S = np.diag(v) @ (xtf[:,1,xl] - 2 * xtf[:,1,:] + xtf[:,1,xr]) / (dx * dx)
    W = np.sin(2 * np.pi * t[k] / 120) * np.sin(np.pi * x / 60) + np.random.normal(0, np.sqrt(tau2_expand), Np*Nx).reshape(Np, Nx)
    xtp[:,0,:] = xtf[:,1,:] + 2 * dt * (A + S + W)
    xtp[:,1,:] = xtf[:,0,:]
    
    # skip filtering step if no observation
    if all(observedT!=k):
        xtf = xtp.copy()
        xtfm[k] = np.mean(xtf[:,0,:], axis=0)
        continue
    
    # log-likelihood at this time point
    dy = yt[k,observedX] - xtp[:,0,observedX]
    llh = -0.5 * (np.log(2 * np.pi * sigma2) + np.sum(dy*dy, axis=1) / sigma2)
    
    # filtering using the residual systematic resampling
    weight = np.exp(llh - np.max(llh)) / np.sum(np.exp(llh - np.max(llh)))
    r = np.random.rand() / Np
    uTemp = u.copy()
    vTemp = v.copy()
    tau2Temp = tau2.copy()
    
    for p in range(Np):
        res[p] = np.floor((weight[p] - r) * Np) + 1
        r += res[p] / Np - weight[p]

    j = 0
    for i in range(Np):
        for _ in range(res[i]):
            xtf[j] = xtp[i]
            u[j] = uTemp[i]
            v[j] = vTemp[i]
            tau2[j] = tau2Temp[i]
            j += 1
    
    xtfm[k] = np.mean(xtf[:,1,:], axis=0)

# prior & posterior distributions of the model parameters
plt.figure(figsize=(10,5))
plt.suptitle('Prior & Posterior Distributions of the Model Parameters')
plt.subplot(2,3,1)
plt.hist(uPrior, bins=20)
plt.xlabel('$u$')
plt.ylabel('$p(u)$')
plt.subplot(2,3,2)
plt.hist(vPrior, bins=20)
plt.xlabel('$v$')
plt.ylabel('$p(v)$')
plt.subplot(2,3,3)
plt.hist(tau2Prior, bins=20)
plt.xlabel('$\\tau^2$')
plt.ylabel('$p(\\tau^2)$')
plt.subplot(2,3,4)
plt.hist(u, bins=20, color='tab:orange')
plt.xlabel('$u$')
plt.ylabel('$p(u|y)$')
plt.subplot(2,3,5)
plt.hist(v, bins=20, color='tab:orange')
plt.xlabel('$v$')
plt.ylabel('$p(v|y)$')
plt.subplot(2,3,6)
plt.hist(tau2, bins=20, color='tab:orange')
plt.xlabel('$\\tau^2$')
plt.ylabel('$p(\\tau^2|y)$')
plt.tight_layout()
plt.show()

# true, simulated, observed & assimilated concentrations
plt.figure(figsize=(8,8))
plt.subplot(2,2,1)
plt.gca().set_title('True')
plt.imshow(yt0, aspect='auto', origin='lower', cmap='autumn')
plt.xticks(np.arange(0, Nx+1, Nx//4), np.arange(0, Nx*dx+1, Nx*dx//4, dtype=int))
plt.xlabel('$x$')
plt.yticks(np.arange(0, Nt+1, Nt//6), np.arange(0, Nt*dt+1, Nt*dt//6, dtype=int))
plt.ylabel('$t$')
plt.subplot(2,2,2)
plt.gca().set_title('Simulation')
plt.imshow(yts, aspect='auto', origin='lower', cmap='autumn')
plt.xticks(np.arange(0, Nx+1, Nx//4), np.arange(0, Nx*dx+1, Nx*dx//4, dtype=int))
plt.xlabel('$x$')
plt.yticks(np.arange(0, Nt+1, Nt//6), np.arange(0, Nt*dt+1, Nt*dt//6, dtype=int))
plt.ylabel('$t$')
plt.subplot(2,2,3)
plt.gca().set_title('Observation')
plt.imshow(yt, aspect='auto', origin='lower', cmap='autumn')
plt.xticks(np.arange(0, Nx+1, Nx//4), np.arange(0, Nx*dx+1, Nx*dx//4, dtype=int))
plt.xlabel('$x$')
plt.yticks(np.arange(0, Nt+1, Nt//6), np.arange(0, Nt*dt+1, Nt*dt//6, dtype=int))
plt.ylabel('$t$')
plt.subplot(2,2,4)
plt.gca().set_title('Assimilation')
plt.imshow(xtfm, aspect='auto', origin='lower', cmap='autumn')
plt.xticks(np.arange(0, Nx+1, Nx//4), np.arange(0, Nx*dx+1, Nx*dx//4, dtype=int))
plt.xlabel('$x$')
plt.yticks(np.arange(0, Nt+1, Nt//6), np.arange(0, Nt*dt+1, Nt*dt//6, dtype=int))
plt.ylabel('$t$')
plt.tight_layout()
plt.show()
